### PR TITLE
AMQP - Expose virtual host attribute

### PR DESCRIPTION
**Pull request by cescoffier**
Related to https://github.com/quarkusio/quarkus/issues/15229

### CODE REVIEW

**Comment by cescoffier** [comment]
FYI  @gemmellr

**Review by kenfinnigan — APPROVED** [review]

**Comment by gemmellr** [comment]
As I said on quarkus/issues/15229, vertx-protons virtualhost client option explicitly sets *both* the TLS SNI value and the AMQP Open frame hostname values to the value provided, overriding the DNS hostname connected to. This PR only exposes usage of the vertx-amqp-client(/vertx-proton) virtualhost option, but seemingly not the related SniServerName option. Doesnt look like it is exposed otherwise? That may be important if in the case being asked about they may not want the SNI to match their broker-internal virtualhost, but some other value, and if so might need to also explicitly configure that too (e.g back to the DNS hostname connected to, or some other common name the server cert uses). 

Usage of vertx-protons virtualhost ClientOption was expected to be rare. For the Open frame the main interaction if wanting something other than the DNS name connected to is use of the setHostname method on the connection object before opening it. As vertx-amqp-client opens the connection itself it obviously cant expose a similar setter, it would have to either manipulate the options given to vetx-proton and introduce its own behaviour, or it might be worth vertx-amqp-client having its own e.g 'amqp-hostname' option to allow it to set the Open hostname using the vertx-proton connection.setHostname(..) method before it then opens it.

Actually, looking at the vertx-amqp-client code, it seems to already both pass the virtualhost option straight through to vertx-proton *and* then use the same option value to explicitly set the hostname on the connection object before opening it, which doesnt really make sense as it should already be done by then. That could be changed to make use of such a new option.

**Comment by cescoffier** [comment]
@gemmellr so we need a new attribute to configure the hostname? 
As you mentioned, the `virtual-host` option is passed as hostname, so if that's not totally correct, what would need to be set?

**Comment by gemmellr** [comment]
Im saying that vertx-amqp-client using the virtualhost option value to call setHostname on the vertx-proton connection should be redundant. Doing that is one of the two things that vertx-proton does with the virtualhost option, and so it should already be done by then. The value is fine, it should just already be set

I'm also saying that reactive-messaging either needs to also expose the SniServerName option, so it can be used to either the SNI value, perhaps to undo what the virtualhost option does to the SNI value. That, or a further option be added to allow solely controlling the AMQP connection hostname (e.g amqp-connection-hostname?),  as that cant be done in isolation with vertx-amqp-client currently (as it lacks the setHostname ability and it cant be added), but can be done with vertx-proton (as it has the setHostname ability).

**Comment by cescoffier** [comment]
Ok, so for the SNI part we would need some fix in the various upstreams.

**Comment by gemmellr** [comment]
Mm, no?

**Comment by cescoffier** [comment]
Oh, I just found the SNI options! So, yes, we could expose it too as `amqp-connection-hostname`. 

So, in the end, we would have 2 new attributes:

* amqp-connection-hostname
* amqp-virtual-host

Both passed, if set, to the AMQP Client Options.

**Comment by gemmellr** [comment]
The `amqp-connection-hostname` was a suggestion for a new option that only controlls the Open hostname (which would indeed need a vertx-amqp-client change to be made). There is already an SniServerName option as youve found for controlling the SNI, just not exposed by reactive-messaging.

The virtualhost option this PR exposes, controls both the Open hostname and the SNI hostname, unless the SNI is already overriden with its option which takes precedence. Using both may be needed for the requesting user (to set SNI back to the DNS value), or the new option above could be added instead.

**Comment by cescoffier** [comment]
Yes, `amqp-connection-hostname` should be `amqp-sni-server-name`

**Comment by gemmellr** [comment]
I dont really follow that. The SNI stuff already has its own explicit option, and is also implicit via the existing virtualhost option. Why wouldnt the explicit SNI option just the existing explicit SNI option?

The `amqp-connection-hostname` suggestion was to add another separate option to vertx-amqp-client and here, that specifically controls only the Open frame hostname value, independent of the SNI. Thats something you can already do with vertx-proton but not with vertx-amqp-client, because it opens the underlying connection before returning it. The new option could be used to set the hostname value [here](https://github.com/vert-x3/vertx-amqp-client/blob/3.9.6/src/main/java/io/vertx/amqp/impl/AmqpConnectionImpl.java#L85-L87), rather than the redundant re-use of the virtualhost option that will already be in effect.

**Comment by cescoffier** [comment]
I don't believe SNI has its own exposed option, at least it's not exposed in the connector.

Here is the configuration code:

```
AmqpClientOptions options = new AmqpClientOptions()
                    .setUsername(username)
                    .setPassword(password)
                    .setHost(host)
                    .setPort(port)
                    .setContainerId(containerId)
                    .setSsl(useSsl)
                    .setReconnectAttempts(reconnectAttempts)
                    .setReconnectInterval(reconnectInterval)
                    .setConnectTimeout(connectTimeout);
```

**Comment by cescoffier** [comment]
I agree with changing the vert.x AMQP client, but it won't be made before the next release.

**Comment by gemmellr** [comment]
It isnt exposed here yet, that was what my original comment on this PR was, that its missing and the user prompting this may need it. It is definitely there in vertx-amqp-client since its there in the options in vertx-proton: https://github.com/vert-x3/vertx-amqp-client/blob/3.9.6/src/main/java/io/vertx/amqp/AmqpClientOptions.java#L499-L502

**Comment by cescoffier** [comment]
Ok, I updated the PR.

It exposes the two options we discussed:

* virtual host
* sni server name

When set, they are passed to the Vert.x AMQP Client Options.

File: smallrye-reactive-messaging-amqp/src/main/java/io/smallrye/reactive/messaging/amqp/AmqpConnector.java [hunk 1]
```diff
@@ -52,6 +52,8 @@
 @ConnectorAttribute(name = "host", direction = INCOMING_AND_OUTGOING, description = "The broker hostname", type = "string", alias = "amqp-host", defaultValue = "localhost")
 @ConnectorAttribute(name = "port", direction = INCOMING_AND_OUTGOING, description = "The broker port", type = "int", alias = "amqp-port", defaultValue = "5672")
 @ConnectorAttribute(name = "use-ssl", direction = INCOMING_AND_OUTGOING, description = "Whether the AMQP connection uses SSL/TLS", type = "boolean", alias = "amqp-use-ssl", defaultValue = "false")
+@ConnectorAttribute(name = "virtual-host", direction = INCOMING_AND_OUTGOING, description = "If set, configure the virtual host name for the connection AMQP Open frame and TLS SNI server name (if TLS is in use)", type = "string", alias = "amqp-virtual-host")
```
Comment (gemmellr) [inline smallrye-reactive-messaging-amqp/src/main/java/io/smallrye/reactive/messaging/amqp/AmqpConnector.java:55]:
```suggestion
@ConnectorAttribute(name = "virtual-host", direction = INCOMING_AND_OUTGOING, description = "If set, configure the hostname value used for the connection AMQP Open frame and TLS SNI server name (if TLS is in use)", type = "string", alias = "amqp-virtual-host")
```

**Review by gemmellr — COMMENTED** [review]
Looks good, one small tweak suggestion for the descriptions (would need a rebuild for the others) to better indicate what its doing.

A test would be good to ensure it is at least affecting the Open hostname value, I'm putting one together.

**Comment by gemmellr** [comment]
I created a test as referenced just above, you can either pick the commit in or I can do a PR after yours lands.

I also raised a PR for vertx-amqp-client to add the new option that could be used in a future update.

**Comment by cescoffier** [comment]
Thanks @gemmellr 

I've applied your suggestion and cherry-picked the test

**Pull request merged**